test/integration/controller_test.go

```diff
@@ -358,6 +358,181 @@ func TestAsyncProvisionWithMultiplePolls(t *testing.T) {
+						Type:   v1beta1.ServiceInstanceConditionReady,
+						Status: v1beta1.ConditionTrue,
+						Reason: "ProvisionedSuccessfully",
+					}); err != nil {
+					t.Fatalf("error waiting for instance to be ready: %v", err)
+				}
+
+				// add a parameter to the instance for the Update()
+				ct.instance.Spec.Parameters = convertParametersIntoRawExtension(t,
+					map[string]interface{}{
+						"param-key": "new-param-value",
+					})
+
+				_, err := ct.client.ServiceInstances(testNamespace).Update(ct.instance)
+				if err != nil {
+					t.Fatalf("error updating instance: %v", err)
+				}
+
+				if err := util.WaitForInstanceCondition(ct.client, ct.instance.Namespace, ct.instance.Name,
+					v1beta1.ServiceInstanceCondition{
+						Type:   v1beta1.ServiceInstanceConditionReady,
+						Status: v1beta1.ConditionFalse,
+						Reason: "UpdatingInstance",
+					}); err != nil {
+					t.Fatalf("error waiting for instance to be updating asynchronously: %v", err)
+				}
+
+				if err := ct.client.ServiceInstances(ct.instance.Namespace).Delete(ct.instance.Name, &metav1.DeleteOptions{}); err != nil {
+					t.Fatalf("failed to delete instance: %v", err)
+				}
+
+				// notify the thread handling DynamicPollLastOperationReaction that it can end the async op
+				done <- true
+
+				if err := util.WaitForInstanceToNotExist(ct.client, ct.instance.Namespace, ct.instance.Name); err != nil {
+					t.Fatalf("error waiting for instance to not exist: %v", err)
+				}
+
+				// We deleted the instance above, clear it so test cleanup doesn't fail
+				ct.instance = nil
+			})
+		})
+	}
+}
+
+// TestServiceInstanceDeleteWithAsyncProvisionInProgress tests that you can
+// delete an instance during an async provision.  Verify the instance is deleted
+// when the provisioning completes regardless of success or failure.
+func TestServiceInstanceDeleteWithAsyncProvisionInProgress(t *testing.T) {
+	cases := []struct {
+		name              string
+		provisionSucceeds bool
+	}{
+		{
+			name:              "provision succeeds",
+			provisionSucceeds: true,
+		},
+		{
+			name:              "provision fails",
+			provisionSucceeds: false,
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			done := make(chan bool)
+
+			ct := controllerTest{
+				t:                            t,
+				broker:                       getTestBroker(),
+				instance:                     getTestInstance(),
+				skipVerifyingInstanceSuccess: true,
+				setup: func(ct *controllerTest) {
+					ct.osbClient.ProvisionReaction.(*fakeosb.ProvisionReaction).Response.Async = true
+					ct.osbClient.PollLastOperationReaction = fakeosb.DynamicPollLastOperationReaction(
+						func(_ *osb.LastOperationRequest) (*osb.LastOperationResponse, error) {
+							state := osb.StateInProgress
+							select {
+							// nonblocking check for a message to finish async provisioning, return inProgress otherwise
+							case <-done:
```

Comment:
There might still be a race here. If a reconciliation starts, and then the deletion timestamp is set, we'd pull `true` out of `done` but then fail to update the instance (as our resource is out of date). The next reconciliation loop, we'd then be back to the channel being empty.

I think a better gauge would be if the channel is closed or not.

Reply:
@kibbles-n-bytes, are you ok with instead moving the declaration and initialization of ```state := osb.StateInProgress```  up to the top of the Run() ?  This way reading from the channel toggles us to a done state and we stay there for any additional polls.

Reply:
What about using an atomic int instead of a channel at all? I think that makes it clearer what is going on from a readability standpoint.